Training24/Test.c

```diff
@@ -1,14 +1,193 @@
+   };
+   int numTests = sizeof (tests) / sizeof (tests[0]);
+   const char** input = malloc (numTests * sizeof (const char*));
+   char** actualResults = malloc (numTests * sizeof (char*));
+   int* pass = malloc (numTests * sizeof (int));
+   if (input == NULL || actualResults == NULL || pass == NULL) {
+      printf ("Memory allocation failed\n");
+      return;
+   }
+   for (int i = 0; i < numTests; ++i) {
+      input[i] = tests[i].input;
+      char* temp = malloc (strlen (tests[i].input) + 1);
+      if (temp == NULL) {
+         printf ("Memory allocation failed for temp\n");
+         return;
+      }
+      FilterChar (temp, tests[i].input);
+      bool result = IsPalindrome (temp);
+      actualResults[i] = result ? "Palindrome" : "Not Palindrome";
+      pass[i] = (strcmp (actualResults[i], tests[i].expected) == 0);
+      free (temp);
+   }
+   // Call the function to format and print the results
+   FormatResult (numTests, input, actualResults, pass);
+   free (input);
+   free (actualResults);
+   free (pass);
+}
+
+/// <summary>Test cases for ReverseNumber function.</summary>
+void TestReverseNumber () {
+   struct {
+      int input, expected;
+   } tests[] = {
+       {1234, 4321}, {0, 0}, {1000, 1},
+       {10, 1}, {9, 9}, {-11234, -43211}, {-42145142, -24154124}
+   };
+   int numTests = sizeof (tests) / sizeof (tests[0]);
+   int* input = malloc (numTests * sizeof (int));
+   int* actualResults = malloc (numTests * sizeof (int));
+   int* pass = malloc (numTests * sizeof (int));
+   if (input == NULL || actualResults == NULL || pass == NULL) {
+      printf ("Memory allocation failed\n");
+      return;
+   }
+   for (int i = 0; i < numTests; ++i) {
+      input[i] = tests[i].input;
+      actualResults[i] = ReverseNumber (tests[i].input);
+      pass[i] = (actualResults[i] == tests[i].expected);
+   }
+   // Call the function to format and print the results
+   FormatResult2 (numTests, input, actualResults, pass);
+   free (input);
+   free (actualResults);
+   free (pass);
+}
+
+void main () {
+   // Run test cases
+   TestPalindrome ();
+   TestReverseNumber ();
+   char choice;
+   while (1) {
+      printf (MAGENTA "Enter the input: " RESET);
+      size_t bufsize = 0;
+      char* buffer = NULL;
+      while (1) {
+         char* temp = realloc (buffer, bufsize + 1);
+         if (temp == NULL) {
+            fprintf (stderr, "Memory allocation failed\n");
+            free (buffer);
+            return;
+         }
+         buffer = temp;
+         int ch = getchar ();
+         if (ch == '\n' || ch == EOF) break;
+         buffer[bufsize++] = (char)ch;
+      }
+      if (bufsize > 4000) printf (RED "The input exceeds the maximum length.\n" RESET);
+      else if (bufsize > 0) {
+         buffer[bufsize] = '\0';
+         // Check for empty string or whitespace
+         if (strlen (buffer) == 0 || strspn (buffer, " \t\r\n") == strlen (buffer)) printf ("Empty string is not a palindrome.\n");

```

Comment:
See [Guard](https://en.wikipedia.org/wiki/Guard_(computer_science)).
```suggestion
         if (strlen (buffer) == 0 || strspn (buffer, " \t\r\n") == strlen (buffer)) {
   printf ("Empty string is not a palindrome.\n"); continue;
}
```